src/libawkward/array/ListArray.cpp

```diff
@@ -812,6 +812,110 @@ namespace awkward {
+          target,
+          starts_.length(),
+          starts_.offset(),
+          stops_.offset()
+        );
+        util::handle_error(err2, classname(), identities_.get());
+
+        Index64 index(tolength);
+        IndexOf<T> starts(starts_.length());
+        IndexOf<T> stops(starts_.length());
+        struct Error err3 = util::awkward_ListArray_rpad_axis1_64<T>(
+          index.ptr().get(),
+          starts_.ptr().get(),
+          stops_.ptr().get(),
+          starts.ptr().get(),
+          stops.ptr().get(),
+          target,
+          starts_.length(),
+          starts_.offset(),
+          stops_.offset());
+        util::handle_error(err3, classname(), identities_.get());
+
+        std::shared_ptr<IndexedOptionArray64> next = std::make_shared<IndexedOptionArray64>(Identities::none(), util::Parameters(), index, content());
+        return std::make_shared<ListArrayOf<T>>(Identities::none(), parameters_, starts, stops, next.get()->simplify());
+      }
+    }
+    else {
+      return std::make_shared<ListArrayOf<T>>(Identities::none(), parameters_, starts_, stops_, content_.get()->rpad(target, axis, depth + 1));
+    }
+  }
+
+  template <typename T>
+  const std::shared_ptr<Content> ListArrayOf<T>::rpad_and_clip(int64_t target, int64_t axis, int64_t depth) const {
+    int64_t toaxis = axis_wrap_if_negative(axis);
+    if (toaxis == depth) {
+      Index64 index(target);
+      struct Error err = awkward_index_rpad_and_clip_axis0_64(
+        index.ptr().get(),
+        target,
+        length());
+      util::handle_error(err, classname(), identities_.get());
+      std::shared_ptr<IndexedOptionArray64> next = std::make_shared<IndexedOptionArray64>(Identities::none(), util::Parameters(), index, shallow_copy());
+      return next.get()->simplify();
+    }
+    else if (toaxis == depth + 1) {
+      Index64 index(starts_.length()*target);
+      Index64 starts(starts_.length());
+      Index64 stops(starts_.length());
+
+      struct Error err1 = awkward_index_rpad_and_clip_axis1_64(
+        starts.ptr().get(),
+        stops.ptr().get(),
+        target,
+        starts.length());
+      util::handle_error(err1, classname(), identities_.get());
+
+      struct Error err2 = util::awkward_ListArray_rpad_and_clip_axis1_64<T>(
+        index.ptr().get(),
+        starts_.ptr().get(),
+        stops_.ptr().get(),
+        target,
+        starts_.length(),
+        starts_.offset(),
+        stops_.offset());
+      util::handle_error(err2, classname(), identities_.get());
+
+      std::shared_ptr<IndexedOptionArray64> next = std::make_shared<IndexedOptionArray64>(Identities::none(), util::Parameters(), index, content());
+      return std::make_shared<ListArray64>(Identities::none(), parameters_, starts, stops, next.get()->simplify());
+    }
+    else {
+      return std::make_shared<ListArrayOf<T>>(Identities::none(), parameters_, starts_, stops_, content_.get()->rpad_and_clip(target, axis, depth + 1));
+    }
+  }
+
```

Comment:
For variable-length lists (ListArray and ListOffsetArray), `rpad` should return a new variable-length list (it does) and `rpad_and_clip` should return a regular-length list (it does not):

```python
    assert str(awkward1.typeof(listoffsetarray.rpad(5, 1))) == "var * ?float64"           # passes
    assert str(awkward1.typeof(listoffsetarray.rpad_and_clip(5, 1))) == "5 * ?float64"    # fails
```

I see that these are two implementations, but the `rpad_and_clip` needs to be reworked to (a) not generate new `starts` and `stops` because they're completely regular, and (b) return a RegularArray.

That's more straightforward for ListOffsetArray, which is guaranteed to have contiguous contents. For ListArray, you'll need a `carry`. Actually, it would be reasonable to start `ListArray::rpad_and_clip` with `toListOffsetArray64()` and let `ListOffsetArray::rpad_and_clip` do the work. That applies to `rpad_and_clip` (because it has to return a RegularArray, which has contiguous contents, and therefore contents need to be carried) and not `rpad` (because it's going to return a ListArray, anyway).

Reply:
@jpivarski - thanks! Shall I make a bugfix PR?

Reply:
You could either update this PR or merge this and start a bugfix PR. In the latter case, be sure to remove the failing tests I added so that master doesn't break, then re-introduce them into your bugfix PR(s). Those failing tests are supposed to be informative, but master is not allowed to break.

If you do want to do this through a bugfix PR, let me know when you've removed the failing tests and all tests pass so that I can merge and deploy (because I like to be in the habit of deploying every change to master).

Reply:
I'll do it in this PR - I thought you've already merged it, but it was a message about something else :-)

Reply:
done